File: Array/151M. Reverse Words in a String.py
# Method: using string.split()
class Solution(object):
    def reverseWords(self, s):
        """
        :type s: str
        :rtype: str
        """
        sl = s.split()
        ans = ""
        for i in range(len(sl)-1, -1, -1):
            ans += sl[i] + " "

        return ans[:-1]

'''
Method 2: Two pointers
'''
class Solution(object):
    def reverseWords(self, s):str
        """
        left, right = 0, 0
        words = []
        while right < len(s):
            if s[right]==' ' and right!=left:
                words.append(s[left:right])
                left = right + 1
            elif s[right]==' ' and s[left]== ' ':
                left += 1
            elif right == len(s)-1:
                words.append(s[left:right+1])
            right += 1
        
        ans = ''
        for i in range(len(words)-1, 0, -1):
            ans = ans + words[i] + ' '
            
        return ans + words[0]

'''
Method 3:
'''
class Solution(object):
    def reverseWords(self, s):
        """
        :type s: str
        :rtype: str
        """
        sub_str = ""
        l = []
        i = 0
        while i < len(s):
            if s[i] != " ":
                sub_str += s[i]
                i += 1
                continue
            else:
                l.append(sub_str+" ")
                sub_str = ""
                while i < len(s) and s[i] == " ":
                    i += 1
        if len(sub_str):
            l.append(sub_str+" ")
        res = ""
        while len(l):
            res += l.pop()
        while res[0] == " ":
            res = res[1:]
        while res[-1] == " ":
            res = res[:-1]
        return res

File: Array/151M. Reverse Words in a String.cpp
/*
151M. Reverse Words in a String.py
String

Description:
Given an input string s, reverse the order of the words.
A word is defined as a sequence of non-space characters. The words in s will be separated by at least one space.
Return a string of the words in reverse order concatenated by a single space.
Note that s may contain leading or trailing spaces or multiple spaces between two words. The returned string should only have a single space separating the words. Do not include any extra spaces.


Example 1:
Input: s = "the sky is blue"
Output: "blue is sky the"

Example 2:
Input: s = "  hello world  "
Output: "world hello"
Explanation: Your reversed string should not contain leading or trailing spaces.

Example 3:
Input: s = "a good   example"
Output: "example good a"
Explanation: You need to reduce multiple spaces between two words to a single space in the reversed string.

Example 4:
Input: s = "  Bob    Loves  Alice   "
Output: "Alice Loves Bob"

Example 5:
Input: s = "Alice does not even like bob"
Output: "bob like even not does Alice"

Similar Question:
Reverse Words in a String II - Medium
*/

/* Method: there are two scenarios that need to be considered:
           1. if s[i] is not a space, then we save it to a string, word;
           2. if s[i] is a space, then we need to check the size of word.
              if word.size() is not 0, it means we have save a complete word and we save it to a vector.
*/
class Solution {
public:
    string reverseWords(string s) {
        vector<string> v;
        int right=0;
        string word="";
        while(right<s.size()){
            if(!isspace(s[right])){
                word += s[right];
                if(right==s.size()-1){
                    v.push_back(word);
                }
            }else{
                if(word.size()!=0){
                    v.push_back(word);
                    word = "";
                }
            }
            right++;
        }

        string ans;
        for(int i=v.size()-1;i>-1;i--){
            ans += v[i] +" ";
        }
        ans.pop_back();
        return ans;
    }
};
